runtime/master/src/main/java/edu/snu/nemo/runtime/master/scheduler/IntermediateDataLocationAwareSchedulingConstraint.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2018 Seoul National University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.snu.nemo.runtime.master.scheduler;
+
+import edu.snu.nemo.common.ir.edge.executionproperty.CommunicationPatternProperty;
+import edu.snu.nemo.common.ir.executionproperty.AssociatedProperty;
+import edu.snu.nemo.common.ir.vertex.executionproperty.IntermediateDataLocationAwareSchedulingProperty;
+import edu.snu.nemo.runtime.common.RuntimeIdGenerator;
+import edu.snu.nemo.runtime.common.plan.StageEdge;
+import edu.snu.nemo.runtime.common.plan.Task;
+import edu.snu.nemo.runtime.master.BlockManagerMaster;
+import edu.snu.nemo.runtime.master.resource.ExecutorRepresenter;
+import org.apache.reef.annotations.audience.DriverSide;
+
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * A scheduling constraint that tries to pick the executor where the intermediate data to read reside.
+ */
+@ThreadSafe
+@DriverSide
+@AssociatedProperty(IntermediateDataLocationAwareSchedulingProperty.class)
+public final class IntermediateDataLocationAwareSchedulingConstraint implements SchedulingConstraint {
```

Comment:
How about reusing the existing SourceLocationAwareSchedulingConstraint? (and renaming the class to DataLocalityConstraint?)

e.g., 
DataLocalityConstraint(task) {
  if (task.source) {
    // source data locality logic
  } else {
    // intermediate data locality logic
  }
}